### PR TITLE
Add Android benchmark

### PR DESCRIPTION
**Pull request by ZacSweers**
This adds a benchmark with the androidx benchmark tool, relates to #17 

Takeaways - we're dealing with microseconds in most cases. There is a significant overhead, but it's so fast already that I'm not terribly concerned. This is also amortized as the chain becomes more complex. At runtime there's hardly any different in e2e testing.

Results with my pixel 3 (will also test on an old moto x for a lower end device).

```
Benchmark - time (ns) - time (ms) - percent increase
```

Milliseconds benchmark there for perspective.

Event throughput: grouped by number of events:

```
1 item (observable)
RxDogTagAndroidPerf.observable1_false                           212ns    0.000ms
RxDogTagAndroidPerf.observable1_true                         13,267ns    0.013ms  6158.02%

1 item (flowable)
RxDogTagAndroidPerf.flowable1_false                             223ns    0.000ms
RxDogTagAndroidPerf.flowable1_true                           13,129ns    0.013ms  5787.44%

1000 items (observable)
RxDogTagAndroidPerf.observable1000_false                     17,596ns   0.018ms
RxDogTagAndroidPerf.observable1000_true                     156,953ns   0.157ms  791.98%

1000 items (flowable)
RxDogTagAndroidPerf.flowable1000_false                       17,854ns   0.018ms
RxDogTagAndroidPerf.flowable1000_true                       143,646ns   0.144ms  704.56%

1000000 items (observable)
RxDogTagAndroidPerf.observable1000000_false              17,078,596ns  17.079ms
RxDogTagAndroidPerf.observable1000000_true              161,099,912ns  161.100ms   843.29%

1000000 items (flowable)
RxDogTagAndroidPerf.flowable1000000_false                17,790,262ns  17.790ms
RxDogTagAndroidPerf.flowable1000000_true                152,887,724ns  152.888ms   759.39%
```

Subscribe cost: grouped by complexity:

```
Simple (observable)
RxDogTagAndroidPerf.observable_false_subscribe_simple           112ns     0.000ms
RxDogTagAndroidPerf.observable_true_subscribe_simple         13,017ns     0.013ms  11522.32%

Simple (flowable)
RxDogTagAndroidPerf.flowable_false_subscribe_simple             147ns    0.000ms
RxDogTagAndroidPerf.flowable_true_subscribe_simple           12,725ns    0.013ms  8556.46%

Complex (observable)
RxDogTagAndroidPerf.observable_false_subscribe_complex        5,322ns   0.005ms
RxDogTagAndroidPerf.observable_true_subscribe_complex        25,046ns   0.025ms  370.61%

Complex (flowable)
RxDogTagAndroidPerf.flowable_false_subscribe_complex          8,376ns   0.008ms
RxDogTagAndroidPerf.flowable_true_subscribe_complex          49,184ns   0.049ms  487.20%
```

E2E amortized cost:

```
Observable
RxDogTagAndroidPerf.observable_false_e2e                     99,010ns  0.099ms
RxDogTagAndroidPerf.observable_true_e2e                     145,833ns  0.146ms  47.29%

Flowable
RxDogTagAndroidPerf.flowable_false_e2e                      126,371ns  0.126ms
RxDogTagAndroidPerf.flowable_true_e2e                       153,107ns  0.153ms  21.16%
```

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.3.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.4.1-all.zip
```
Comment (dlew) [inline gradle/wrapper/gradle-wrapper.properties]: Did you run the new wrapper call? I'm thought more changed than just the ZIP between 5.3 and 5.4.

File: android-benchmark/benchmark/src/main/kotlin/com/uber/rxdogtag/androidbenchmark/DataParser.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    grouping to results.filter {
+      grouping.matchFunction(it.benchmark)
+    }
+  }
+  val benchmarkLength = results.maxBy { it.benchmark.length }!!.benchmark.length
+  val scoreLength = results.maxBy { it.formattedScore.length }!!.formattedScore.length
+
+//  check(groupedResults.values.flatten().size == results.size) {
+//    "Missing information!"
+//  }
```
Comment (dlew) [inline android-benchmark/benchmark/src/main/kotlin/com/uber/rxdogtag/androidbenchmark/DataParser.kt]: Nit: Commented-out code

File: settings.gradle [hunk 3]
```diff
@@ -39,5 +39,8 @@ pluginManagement {
 }
 
 rootProject.name = 'rxdogtag-root'
+if (System.getenv("ANDROID_HOME") != null) {
```
Comment (dlew) [inline settings.gradle]: Similar to the JMH benchmarks, maybe we should include a little snippet on how to run the Android benchmarks somewhere?

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (ZacSweers) [reply gradle/wrapper/gradle-wrapper.properties]: I thought so as well and just re-ran, but no new changes.

```bash
./gradlew wrapper --gradle-version 5.4.1 --distribution-type all --no-daemon --stacktrace
```

See anything that looks off?

File: settings.gradle (hunk 3 above)
Comment (ZacSweers) [reply settings.gradle]: Good call, added here - b1c832c

Will flesh out a full wiki after both of these are in as well

File: android-benchmark/benchmark/src/main/kotlin/com/uber/rxdogtag/androidbenchmark/DataParser.kt (hunk 2 above)
Comment (ZacSweers) [reply android-benchmark/benchmark/src/main/kotlin/com/uber/rxdogtag/androidbenchmark/DataParser.kt]: derp b1c832c

**Comment by ZacSweers** [comment]
Results on a lower end device (2013 moto x running kitkat)

Event throughput: grouped by number of events:

```
1 item (observable)
RxDogTagAndroidPerf.observable1_false                           3,748ns   0.004ms
RxDogTagAndroidPerf.observable1_true                           36,069ns   0.036ms  862.35%

1 item (flowable)
RxDogTagAndroidPerf.flowable1_false                             4,695ns   0.005ms
RxDogTagAndroidPerf.flowable1_true                             30,520ns   0.031ms  550.05%

1000 items (observable)
RxDogTagAndroidPerf.observable1000_false                      213,643ns    0.214ms
RxDogTagAndroidPerf.observable1000_true                     2,380,589ns    2.381ms  1014.28%

1000 items (flowable)
RxDogTagAndroidPerf.flowable1000_false                        183,122ns    0.183ms
RxDogTagAndroidPerf.flowable1000_true                       2,289,027ns    2.289ms  1150.00%

1000000 items (observable)
RxDogTagAndroidPerf.observable1000000_false               210,163,283ns  210.163ms
RxDogTagAndroidPerf.observable1000000_true              3,919,822,979ns  3919.823ms   1765.13%

1000000 items (flowable)
RxDogTagAndroidPerf.flowable1000000_false                 344,147,719ns  344.148ms
RxDogTagAndroidPerf.flowable1000000_true                3,949,153,058ns  3949.153ms   1047.52%
```

Subscribe cost: grouped by complexity:

```
Simple (observable)
RxDogTagAndroidPerf.observable_false_subscribe_simple           1,312ns    0.001ms
RxDogTagAndroidPerf.observable_true_subscribe_simple           20,982ns    0.021ms  1499.24%

Simple (flowable)
RxDogTagAndroidPerf.flowable_false_subscribe_simple             4,108ns   0.004ms
RxDogTagAndroidPerf.flowable_true_subscribe_simple             26,705ns   0.027ms  550.07%

Complex (observable)
RxDogTagAndroidPerf.observable_false_subscribe_complex         45,780ns   0.046ms
RxDogTagAndroidPerf.observable_true_subscribe_complex         111,908ns   0.112ms  144.45%

Complex (flowable)
RxDogTagAndroidPerf.flowable_false_subscribe_complex          111,908ns  0.112ms
RxDogTagAndroidPerf.flowable_true_subscribe_complex           183,122ns  0.183ms  63.64%
```

E2E amortized cost:

```
Observable
RxDogTagAndroidPerf.observable_false_e2e                      457,805ns  0.458ms
RxDogTagAndroidPerf.observable_true_e2e                       549,366ns  0.549ms  20.00%

Flowable
RxDogTagAndroidPerf.flowable_false_e2e                        549,366ns  0.549ms
RxDogTagAndroidPerf.flowable_true_e2e                         579,887ns  0.580ms   5.56%
```

File: android-benchmark/benchmark/src/androidTest/kotlin/com/uber/rxdogtag/androidbenchmark/RxDogTagAndroidPerf.kt [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+  fun observable1000_false() {
+    val observable = observableInstance(1000)
+    benchmarkRule.measureRepeated {
+      observable.subscribe()
+    }
+  }
+
+  @Test
+  fun flowable1000000_false() {
+    val flowable = flowableInstance(1000000)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply android-benchmark/benchmark/src/androidTest/kotlin/com/uber/rxdogtag/androidbenchmark/RxDogTagAndroidPerf.kt]: I always forget you can do this cc3faf8

**Comment by neworld** [comment]
I was wondering, that is the cost of overhead if all subscribers has error handler. I ran tests on `One Plus 5T`. I found impact is almost non existing:

```
1 item (observable)
RxDogTagAndroidPerf.observable1_false                                147ns    0.000ms
RxDogTagAndroidPerf.observable1_true_withErrorHandler                377ns    0.000ms  156.46%
RxDogTagAndroidPerf.observable1_true                               8,880ns    0.009ms  5940.82%

1 item (flowable)
RxDogTagAndroidPerf.flowable1_false                                  166ns    0.000ms
RxDogTagAndroidPerf.flowable1_true_withErrorHandler                  395ns    0.000ms  137.95%
RxDogTagAndroidPerf.flowable1_true                                 9,208ns    0.009ms  5446.99%

1000 items (observable)
RxDogTagAndroidPerf.observable1000_false                          14,857ns   0.015ms
RxDogTagAndroidPerf.observable1000_true_withErrorHandler          21,211ns   0.021ms  42.77%
RxDogTagAndroidPerf.observable1000_true                          107,109ns   0.107ms  620.93%

1000 items (flowable)
RxDogTagAndroidPerf.flowable1000_false                            16,300ns   0.016ms
RxDogTagAndroidPerf.flowable1000_true_withErrorHandler            19,704ns   0.020ms  20.88%
RxDogTagAndroidPerf.flowable1000_true                            109,097ns   0.109ms  569.31%

1000000 items (observable)
RxDogTagAndroidPerf.observable1000000_false                   14,103,489ns  14.103ms
RxDogTagAndroidPerf.observable1000000_true_withErrorHandler   16,216,302ns  16.216ms    14.98%
RxDogTagAndroidPerf.observable1000000_true                   121,664,374ns  121.664ms   762.65%

1000000 items (flowable)
RxDogTagAndroidPerf.flowable1000000_false                     16,571,563ns  16.572ms
RxDogTagAndroidPerf.flowable1000000_true_withErrorHandler     19,984,531ns  19.985ms    20.60%
RxDogTagAndroidPerf.flowable1000000_true                     117,842,865ns  117.843ms   611.11%
```

Also, I found this report is using min values. Meanwhile, the html report is using means and they are not small:

image

Any ideas why such big different?

**Comment by ShaishavGandhi** [comment]
That's expected since we only decorate with a DogTagObserver if the subscriber doesn't have custom error handling. If you're already defining an error handler, you're probably already thinking about logging your error.

**Pull request merged**